sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala

```diff
@@ -25,18 +25,12 @@ import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.catalyst.trees.TreePattern._
 
 /**
- * A helper class used to detect duplicate relations fast in `DeduplicateRelations`
+ * A helper class used to detect duplicate relations fast in `DeduplicateRelations`. Two relations
+ * are duplicated if:
+ *  1. they are the same class.
+ *  2. they have the same output attribute IDs.
```

Comment:
Could two relations be of different classes but have same output attributes IDs?

Reply:
If they are different classes, but have same output attributes IDs, then they are not duplicated? I guess the issue of duplicated relations will happen on them too no matter they are same or not class, no?

Reply:
CTE ref node has the same output attribute IDs as the CTE def node, and they are not duplicated. Maybe I should put it in the document as well. Otherwise, I can't think of any case that 2 relations of different classes have the same output attribute IDs.

Reply:
Yea, I think it's better to put this case into the document.